Assets/MixedRealityToolkit/Services/BaseEventSystem.cs

```diff
@@ -13,56 +13,209 @@ namespace Microsoft.MixedReality.Toolkit
     /// </summary>
     public abstract class BaseEventSystem : BaseService, IMixedRealityEventSystem
     {
-        #region IMixedRealityEventSystem Implementation
-
         private static int eventExecutionDepth = 0;
-        private readonly WaitUntil doneExecutingEvents = new WaitUntil(() => eventExecutionDepth == 0);
+        private readonly Type eventSystemHandlerType = typeof(IEventSystemHandler);
+
+        private List<(Type, IEventSystemHandler)> handlersToAdd = new List<(Type, IEventSystemHandler)>();
+        private List<(Type, IEventSystemHandler)> handlersToRemove = new List<(Type, IEventSystemHandler)>();
+
+        #region IMixedRealityEventSystem Implementation
 
         /// <inheritdoc />
         public List<GameObject> EventListeners { get; } = new List<GameObject>();
 
+        /// <inheritdoc />
+        public Dictionary<Type, HashSet<IEventSystemHandler>> HandlerEventListeners { get; } = new Dictionary<Type, HashSet<IEventSystemHandler>>();
+
         /// <inheritdoc />
         public virtual void HandleEvent<T>(BaseEventData eventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
         {
             Debug.Assert(!eventData.used);
+
+            HashSet<IEventSystemHandler> handlers;
+            if (!HandlerEventListeners.TryGetValue(typeof(T), out handlers))
+            {
+                return;
+            }
+
             eventExecutionDepth++;
 
-            for (int i = EventListeners.Count - 1; i >= 0; i--)
+            foreach (IEventSystemHandler handler in handlers)
             {
-                ExecuteEvents.Execute(EventListeners[i], eventData, eventHandler);
+                Debug.Assert(handler is T);
+                eventHandler.Invoke((T)handler, eventData);
             }
 
             eventExecutionDepth--;
+
+            if(eventExecutionDepth == 0 && (handlersToRemove.Count > 0 || handlersToAdd.Count > 0))
```

Comment:
I like this better than making it asynchronous but in its current form it does not preserve execution order. E.g. if a handler gets added and then removed while handling this event, we won't produce the right result. I think we can fix that having a single list for add/remove.